File: C++/concurrency.cpp
#include <iostream>
#include <thread>
#include <atomic>
#include <vector>
#include <chrono>
#include <future>

using namespace std;
using namespace chrono;

constexpr int NUM_THREADS = 8;
constexpr long OPS_PER_THREAD = 10000000;

atomic<long> counter{0};

void atomic_increment() {
    for (long i = 0; i < OPS_PER_THREAD; i++) {
        counter.fetch_add(1);
    }
}

long parallel_sum(long start, long end) {
    long sum = 0;
    for (long i = start; i < end; i++) {
        sum += i * i;
    }
    return sum;
}

long thread_pool_task(long start) {
    long sum = 0;
    for (long i = start; i < start + OPS_PER_THREAD; i++) {
        sum += i * i;
    }
    return sum;
}

int main() {
    cout << "C++ Concurrency Benchmark (" << NUM_THREADS << " threads, "
         << OPS_PER_THREAD << " ops/thread)\n";
    
    long total_ops = (long)NUM_THREADS * OPS_PER_THREAD;
    auto t0 = high_resolution_clock::now();
    vector<thread> threads;
    for (int i = 0; i < NUM_THREADS; i++) threads.emplace_back(atomic_increment);
    for (auto& t : threads) t.join();
    auto t1 = high_resolution_clock::now();
    cout << "Atomic: " << duration<double, milli>(t1 - t0).count() << " ms ("
         << (total_ops / duration<double>(t1 - t0).count()) << " ops/sec)\n";
    
    t0 = high_resolution_clock::now();
    vector<future<long>> futures;
    for (int i = 0; i < NUM_THREADS; i++) {
        futures.push_back(async(launch::async, parallel_sum, i * OPS_PER_THREAD, (i + 1) * OPS_PER_THREAD));
    }
    long sum = 0;
    for (auto& f : futures) sum += f.get();
    t1 = high_resolution_clock::now();
    cout << "Parallel sum: " << duration<double, milli>(t1 - t0).count() << " ms ("
         << (total_ops / duration<double>(t1 - t0).count()) << " ops/sec)\n";
    
    threads.clear();
    t0 = high_resolution_clock::now();
    for (int i = 0; i < NUM_THREADS; i++) {
        threads.emplace_back(thread_pool_task, i * OPS_PER_THREAD);
    }
    for (auto& t : threads) t.join();
    t1 = high_resolution_clock::now();
    cout << "Thread pool: " << duration<double, milli>(t1 - t0).count() << " ms ("
         << (total_ops / duration<double>(t1 - t0).count()) << " ops/sec)\n";
    
    return 0;
}

File: Rust/concurrency.rs
use std::thread;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;

const NUM_THREADS: usize = 8;
const OPS_PER_THREAD: usize = 10_000_000;

fn main() {
    println!("Rust Concurrency Benchmark ({} threads, {} ops/thread)",
        NUM_THREADS, OPS_PER_THREAD);
    
    let counter = Arc::new(AtomicU64::new(0));
    let t0 = std::time::Instant::now();
    let handles: Vec<_> = (0..NUM_THREADS)
        .map(|_| {
            let c = counter.clone();
            thread::spawn(move || {
                for _ in 0..OPS_PER_THREAD {
                    c.fetch_add(1, Ordering::Relaxed);
                }
            })
        })
        .collect();
    for h in handles { h.join().unwrap(); }
    let t1 = std::time::Instant::now();
    let total_ops = NUM_THREADS * OPS_PER_THREAD;
    println!("Atomic: {:.2} ms ({:.0} ops/sec)",
        (t1 - t0).as_secs_f64() * 1000.0,
        total_ops as f64 / (t1 - t0).as_secs_f64());
    
    let t0 = std::time::Instant::now();
    let handles: Vec<_> = (0..NUM_THREADS)
        .map(|i| {
            let per_thread = OPS_PER_THREAD;
            thread::spawn(move || {
                let start = i * per_thread;
                let end = start + per_thread;
                let mut sum: u64 = 0;
                for j in start..end {
                    sum += (j as u64) * (j as u64);
                }
                sum
            })
        })
        .collect();
    let _sum: u64 = handles.into_iter().map(|h| h.join().unwrap()).sum();
    let t1 = std::time::Instant::now();
    println!("Parallel sum: {:.2} ms ({:.0} ops/sec)",
        (t1 - t0).as_secs_f64() * 1000.0,
        total_ops as f64 / (t1 - t0).as_secs_f64());
    
    let t0 = std::time::Instant::now();
    let handles: Vec<_> = (0..NUM_THREADS)
        .map(|i| {
            thread::spawn(move || {
                let start = i * OPS_PER_THREAD;
                let mut sum: u64 = 0;
                for j in start..start + OPS_PER_THREAD {
                    sum += (j as u64) * (j as u64);
                }
                sum
            })
        })
        .collect();
    let _sum: u64 = handles.into_iter().map(|h| h.join().unwrap()).sum();
    let t1 = std::time::Instant::now();
    println!("Thread pool: {:.2} ms ({:.0} ops/sec)",
        (t1 - t0).as_secs_f64() * 1000.0,
        total_ops as f64 / (t1 - t0).as_secs_f64());
}
